evidence/reactor.go

```diff
@@ -21,233 +46,346 @@ const (
+	defer r.evidenceCh.Close()
+
+	for {
+		select {
+		case envelope := <-r.evidenceCh.In():
+			if err := r.handleMessage(r.evidenceCh.ID(), envelope); err != nil {
+				r.Logger.Error("failed to process envelope", "ch_id", r.evidenceCh.ID(), "envelope", envelope, "err", err)
+				r.evidenceCh.Error() <- p2p.PeerError{
+					PeerID:   envelope.From,
+					Err:      err,
+					Severity: p2p.PeerErrorSeverityLow,
+				}
+			}
+
+		case <-r.closeCh:
+			r.Logger.Debug("stopped listening on evidence channel; closing...")
 			return
-		case nil:
-		default:
-			// continue to the next piece of evidence
-			evR.Logger.Error("Evidence has not been added", "evidence", evis, "err", err)
 		}
 	}
 }
 
-// SetEventBus implements events.Eventable.
-func (evR *Reactor) SetEventBus(b *types.EventBus) {
-	evR.eventBus = b
+// processPeerUpdate processes a PeerUpdate, returning an error upon failing to
+// handle the PeerUpdate or if a panic is recovered. For new or live peers it
+// will check if an evidence broadcasting goroutine needs to be started. For
+// down or removed peers, it will check if an evidence broadcasting goroutine
+// exists and signal that it should exit.
+//
+// FIXME: The peer may be behind in which case it would simply ignore the
+// evidence and treat it as invalid. This would cause the peer to disconnect.
+// The peer may also receive the same piece of evidence multiple times if it
+// connects/disconnects frequently from the broadcasting peer(s).
+//
+// REF: https://github.com/tendermint/tendermint/issues/4727
+func (r *Reactor) processPeerUpdate(peerUpdate p2p.PeerUpdate) (err error) {
+	defer func() {
+		if e := recover(); e != nil {
+			err = fmt.Errorf("panic in processing peer update: %v", e)
+		}
+	}()
+
+	r.Logger.Debug("received peer update", "peer", peerUpdate.PeerID.String(), "status", peerUpdate.Status)
+
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+
+	switch peerUpdate.Status {
+	case p2p.PeerStatusNew, p2p.PeerStatusUp:
+		// Check if we've already started a goroutine for this peer, if not we create
+		// a new done channel so we can explicitly close the goroutine if the peer
+		// is later removed, we increment the waitgroup so the reactor can stop
+		// safely, and finally start the goroutine to broadcast evidence to that peer.
+		_, ok := r.peerRoutines[peerUpdate.PeerID.String()]
+		if !ok {
+			r.peerRoutines[peerUpdate.PeerID.String()] = make(chan struct{})
+			r.peerWG.Add(1)
+			go r.broadcastEvidenceLoop(peerUpdate.PeerID, r.peerRoutines[peerUpdate.PeerID.String()])
+		}
+
+	case p2p.PeerStatusDown, p2p.PeerStatusRemoved, p2p.PeerStatusBanned:
+		// Check if we've started an evidence broadcasting goroutine for this peer.
+		// If we have, we signal to terminate the goroutine via the channel's closure.
+		// This will internally decrement the peer waitgroup and remove the peer
+		// from the map of peer evidence broadcasting goroutines.
+		doneCh, ok := r.peerRoutines[peerUpdate.PeerID.String()]
```

Comment:
This can end up closing an already closed channel, and thus panic. This happens e.g. if we receive another `PeerStatusDown` before the goroutine terminates (if the peer is strobing up/down or we receive duplicate updates).

One option would be to involve a `sync.Once` somehow. Another option would be to remove the entry here instead of during `removePeerRoutine`, in which case it's possible for multiple goroutines to briefly be running per peer (if we get another `Up` status before the previous goroutine terminates).

Reply:
Yeah, sigh. Good catch. I wanted to explicitly remove it from the map in `broadcastEvidenceLoop` because that is what is responsible for message processing and handling the channel closure signal, otherwise, like you said we could have multiple running which is what I wanted to avoid. 

Let me think on this a bit...

Reply:
I ended up opting for a simple modification to the current design -- `closer`, which simply wraps a `sync.Once` that ensures we close the channel only once. I've also added a test case for this.

I can imagine we might end up needing `closer` in other places, but for now I'll keep this private to the `evidence` pkg.

Reply:
Yeah, this pattern keeps popping up throughout the P2P stack, we may want a convenience wrapper for it somewhere. I wonder if we can maybe make use of a [`sync.Cond`](https://golang.org/pkg/sync/#Cond) for this kind of synchronization -- although I suppose they don't compose with channel selects.